### ISSUE TITLE
Panache - Use collation in find

### ISSUE THREAD

**Issue by overheadhunter** [quarkusio/quarkus#31254]
### Description

It is not possible to define a collation when executing a find with the ~MongoDB~ PanacheQuery. We want to implement filtering and sorting and apply a collation to be able to enable e.g. find case insensitive.

### Implementation ideas

Same as @FroMage and @loicmathieu did for #9081. But please don't limit it to a single database, as it is useful in any db that supports collation, especially when an index is known to exist for a certain collation and we don't want massive slowdowns for using the connection default collation for which no index exists.

**Comment by loicmathieu** [comment]
@overheadhunter 

I don't understand what you asking for.

> But please don't limit it to a single database, as it is useful in any db that supports collation

A collation is not on a Mongo database but on a Mongo query.

**Comment by overheadhunter** [comment]
@loicmathieu I am asking for collation support in Panache queries. The references issue was mongo-only. It does not solve anything for e.g. Postgres or MariaDB. Instead it should be available in the general Panache API

**Comment by loicmathieu** [comment]
OK, so you are asking for collation support in Hibernate ORM with Panache for databases that supports them.
I'll update the description to be more clear about that.

**Comment by lucamolteni** [comment]
Given a focus on Panache Next / Quarkus Data I think we might close this WDYT @FroMage @yrodiere ?

**Comment by yrodiere** [comment]
> Given a focus on Panache Next / Quarkus Data I think we might close this WDYT [@FroMage](https://github.com/FroMage) [@yrodiere](https://github.com/yrodiere) ?

If you can tell me how the use case can be addressed in Panache Next / Quarkus Data, or that this feature is no longer relevant... sure.

**Comment by FroMage** [comment]
I recently switched to JD `Sort` types in Panache Next, so we already support sorting case insensitive now (which we don't in Panache 1). But I am unsure if that is what the original requester wanted.

I looked at the PR mentionned for Mongo/Panache and it's unclear what "collation" does here, because it applies to the entire query. But the tests only test the sorting. So I'm unsure if it also applies to the query part (say `foo = 'aaa'`) or just the sorting, in which case yes we support specifying the casing strategy for sorting, but not on the query, on the `Sort` options.

Now, I also noticed that in Mongo/Panache the collation is more than just case-sensitivity because you can also specify the language rules for that case sensitivity. Which we can't in Panache Next / Jakarta Data. Perhaps we should bring that to the JD spec as enhancement?

**Comment by overheadhunter** [comment]
> collation is more than just case-sensitivity

exactly, it is also about sorting of non-ascii letters as in [this MySQL example](https://dev.mysql.com/doc/refman/5.7/en/charset-collation-effect.html):

| latin1_swedish_ci | latin1_german1_ci |
|-------------------|-------------------|
| Muffler           | Muffler           |
| MX Systems        | Müller            |
| Müller            | MX Systems        |
| MySQL             | MySQL             |

**Comment by FroMage** [comment]
@overheadhunter do you mind opening an issue at https://github.com/jakartaee/data/issues explaining your use-case?

**Comment by overheadhunter** [comment]
@FroMage forgive me the stupid question, but isn't Panache using Hibernate's vendor-specific capabilities rather than JPA?

[Hibernate supports this since at least version 6](https://docs.hibernate.org/orm/6.0/userguide/html_single/#_collate).

**Comment by yrodiere** [comment]
This is HQL syntax; I thought you were after a higher-level syntax (methods, ...) like in #9081?

Otherwise this is wrong: 

> Today, it is not possible to define a collation when executing a query with the Hiberate ORM with Panache extension as we can in the MongoDB with Panache extension.

Because you can just pass HQL to Panache using `collate()`, and it will be forwarded to Hibernate ORM. In Panache Next / Quarkus Data this will be:

```
@Query("where title like collate(title as ucs_basic)")
List<Book> books(String title);
```


If you *are* after a higher-level construct, then the way to go in the next version of Panache would be something similar to `@Pattern` as mentioned [there](https://docs.hibernate.org/orm/7.3/repositories/html_single/#find-method):

```java
@Find
List<Book> books(@Pattern String title);
```

I guess we could introduce something like:

```java
@Find
List<Book> books(@Collate("ucs_basic") String title);
```

But this opens a can of worm, with questions such as: what about other HQL functions? What if I want `collate` to be applied for ordering? Etc.

So, yeah. Best submit the use case to Jakarta Data and see what the experts think of it. Though I strongly suspect they'll just say "this is too advanced for Data, you should try in Persistence or Hibernate instead". It'll be an interesting journey :)

**Comment by overheadhunter** [comment]
> This is HQL syntax; I thought you were after a higher-level syntax (methods, ...) like in [#9081](https://github.com/quarkusio/quarkus/issues/9081)?

Yes, this was my original intend. I assumed it would be reasonable to have some form of feature parity between Hibernate and MongoDB.

I (falsely?) assumed that the high-level API internally translates to HQL.

> Because you can just pass HQL to Panache using `collate()`, and it will be forwarded to Hibernate ORM.

Something comparable has in fact been my workaround since. In most cases we use `@NamedQuery` or even `@NamedNativeQuery` nowadays, as we concluded that the high-level "one size fits all" solution only works for simple CRUD use cases.

---

My original issue is now more than three years old and I can see that you're steering towards "Panache Next", which I assume is a major refactor. So maybe the idea to have some kind of parity between Hibernate and MongoDB is obsolete. I'm happy with the workaround, so feel free to close this as not planned.

**Comment by FroMage** [comment]
OK, so there's HQL support for this, in comparison expressions. So that's one side of it that works. The other side as @yrodiere mentionned is:

```java
@Find
List<Book> books(@Collate("ucs_basic") String title);
```

And the original side of it is via the `Order/Sort` API:

```
@Find
List<Book> books(Order order);
…
repository.books(Order.by(Sort.descIgnoreCase("title", ...somethingLikeLocale?...));
```

I will open an issue upstream and we'll see where this leads.

**Comment by FroMage** [comment]
There we go: https://github.com/jakartaee/data/issues/1445
